<img src="https://userweb.fct.unl.pt/~jmc.xavier/MSII/iLogos/logo_novafct.png" style="width: 30%;">

# Departamento de Engenharia Mecânica e Industrial
## Mecânica dos Sólidos II

## Deslocamentos transversais de vigas

### Problema 2

A figura representa uma viga ABCDE onde é aplicado, no ponto B, um momento $M = 4$ kNm e entre C e E uma carga distribuída definida por uma função $Q(x)$ de segundo grau, cuja intensidade, nas secções C, D e E é respetivamente $Q_C$ = 10 kN/m , $Q_D$ = 14 kN/m e $Q_E$ = 26 kN/m.  A viga é construída com um perfil cuja rigidez de flexão é $EI= 7.38\times 10^6$ Nm2. Determine:

a) As reações nos apoios da viga.

b) A equação da linha elástica.

c) A flecha no ponto E.

<img src="https://userweb.fct.unl.pt/~jmc.xavier/MSII/Notebooks/Au09/P2/MSII_Au09_P2.png"
style="width: 30%;"/>

In [1]:
import numpy as np
import sympy as sy
from sympy.solvers import solve
import matplotlib.pyplot as plt
import os

cor = '1'
if cor == '1':
    plt.rcParams['axes.facecolor'] = (.15, .15, .15)
    plt.rcParams['figure.facecolor'] = (.15, .15, .15)
    plt.rcParams['font.family'] = 'monospace'
    plt.rcParams['font.size'] = 18
    # plt.rcParams['text.usetex'] = True
    params = {"ytick.color" : (.8, .8, .8),
              "xtick.color" : (.8, .8, .8),
              "grid.color" : (.2, .2, .2),
              "text.color" : (.7, .7, .7),
              "axes.labelcolor" : (.8, .8, .8),
              "axes.edgecolor" : (.15, .15, .15)}
else:
    plt.rcParams['axes.facecolor'] = (.7, .7, .7)
    plt.rcParams['figure.facecolor'] = (.7, .7, .7)
    plt.rcParams['font.family'] = 'monospace'
    plt.rcParams['font.size'] = 18
    # plt.rcParams['text.usetex'] = True
    params = {"ytick.color" : (.1, .1, .1),
              "xtick.color" : (.1, .1, .1),
              "grid.color" : (.2, .2, .2),
              "text.color" : (.1, .1, .1),
              "axes.labelcolor" : (.1, .1, .1),
              "axes.edgecolor" : (.15, .15, .15)}
plt.rcParams.update(params)

def printsymSF(expr):
    expr2 = sy.simplify(sy.factor(expr))
    return expr2.xreplace({n : round(n, 3) for n in expr2.atoms(sy.Number)})

# data structure, units: N, mm, MPa

class varin: pass

d = varin()
W250 = varin()

d.LAB = 1.  # unit: m
d.LBC = 1.  # unit: m
d.LCD = 1.  # unit: m
d.LDE = 1.  # unit: m
d.MB = 4.e3  # unit: N.m
d.QC = 10.e3 # unit: N/m
d.QD = 14.e3 # unit: N/m
d.QE = 26.e3 # unit: N/m

d.LCE = d.LCD + d.LDE
d.LAC = d.LAB + d.LBC
d.LAD = d.LAC + d.LCD
d.LAE = d.LAC + d.LCE

d.EI = 7.38e6 # N.m²

## Resolução

- 0 ::: Formulário

<img src="https://userweb.fct.unl.pt/~jmc.xavier/MSII/Notebooks/Au09/P2/MSII_Au09_P2_2.png"
width="900"/>

- 1 ::: cálculo das reações

\begin{equation*}
\sum F_x = 0
\quad\wedge\quad
\sum F_y = 0
\quad\wedge\quad
\sum M_D = 0
\end{equation*}

No Diagrama de Corpo Livre da viga temos interesse, pelo formulário dos esforços descritos por funções de singularidade, que a carga distribuida parabólica seja obtida pelo princípio da sobreposição de uma carga uniforme ($Q_C$) e por uma carga parabólica ($q(x)$) tal que em $x_C$ se obtem $q(x_C) = 0$.

\begin{equation*}
Q(x) = Q_C^{\textrm{uniforme}} + q(x)^{\textrm{parábola}}
\end{equation*}

Desenvolvendo a função que descreve $q(x)$, num referencial local com origem em $x_C$:

\begin{equation*}
q(x) = Ax² + Bx + C~,
\quad x \in [0,x_{CE}]
\end{equation*}

com,

\begin{equation*}
q(0) = Q_C-Q_C~;~q(x_D) = Q_D-Q_C~;~q(x_E) = Q_E-Q_C
\end{equation*}

In [2]:
a, b, c, x = sy.symbols('a b c x')

def qquad(a, b, c, x):
    return a*x**2 + b*x + c

print(':: referencial local ::')
eq1 = qquad(a, b, c, 0) - (d.QC-d.QC)
print(f'eq1 = {eq1}')
eq2 = qquad(a, b, c, d.LCD) - (d.QD-d.QC)
print(f'eq2 = {eq2}')
eq3 = qquad(a, b, c, d.LCE) - (d.QE-d.QC)
print(f'eq3 = {eq3}')

sol = solve((eq1,eq2,eq3),(a,b,c))
Al, Bl, Cl = sol[a], sol[b], sol[c]
print(f'A(l) = {Al:.1f} | B(l) = {Bl:.1f} | C(l) = {Cl:.1f}')
cargaQCElocal = qquad(Al, Bl, Cl, x)
print(f'\ncargaQCElocal = {cargaQCElocal}')
print(f'\nVerificação: \n x= 0   [m] : QC(l) = {qquad(Al, Bl, Cl, 0)} [N]')
print(f'x= LCD [m] : QD(l) = {qquad(Al, Bl, Cl, d.LCD):.1f} [N]')
print(f'x= LCE [m] : QE(l) = {qquad(Al, Bl, Cl, d.LCE):.1f} [N]')


:: referencial local ::
eq1 = c
eq2 = 1.0*a + 1.0*b + c - 4000.0
eq3 = 4.0*a + 2.0*b + c - 16000.0
A(l) = 4000.0 | B(l) = 0.0 | C(l) = 0.0

cargaQCElocal = 4000.0*x**2

Verificação: 
 x= 0   [m] : QC(l) = 0 [N]
x= LCD [m] : QD(l) = 4000.0 [N]
x= LCE [m] : QE(l) = 16000.0 [N]


- Translação de coordeanda: do referencial local para o referencial global

$$
x \rightarrow x-L_{AC}
\quad\therefore\quad
w(x) = m <x-a>^n
$$

$$
w(x) \equiv Q_{CE}(x) = 4000 (x-L_{AC})²
= 4000 <x-2>^2 \quad\textrm{[N]}
$$


In [3]:
print(':: referencial global ::')
cargaQCEglobal = sy.simplify(qquad(Al, Bl, Cl, x-d.LAC))
print(f'cargaQCEglobal = {cargaQCEglobal} = 4000*(x - 2)**2 [N] ')

:: referencial global ::
cargaQCEglobal = 16000.0*(0.5*x - 1)**2 = 4000*(x - 2)**2 [N] 


- Resultante das cargas distribuidas

In [4]:
print(f'carregamento uniforme:')
d.Quni_eq = d.QC*d.LCE
d.Quni_xcent = d.LAC+d.LCE/2
print(f'd.Quni_eq = {d.Quni_eq} [N]')
print(f'd.Quni_xcent = {d.Quni_xcent} [m]')

print(f'carregamento parabólico:')
d.Qparab_eq = sy.integrate(cargaQCElocal,(x,0,d.LCE))
d.Qparab_xcent = d.LAC+sy.integrate(cargaQCElocal*x,(x,0,d.LCE))/d.Qparab_eq
print(f'd.Qparab_eq = {d.Qparab_eq:.1f} [N]')
print(f'd.Qparab_xcent = {d.Qparab_xcent:.1f} [m]')

carregamento uniforme:
d.Quni_eq = 20000.0 [N]
d.Quni_xcent = 3.0 [m]
carregamento parabólico:
d.Qparab_eq = 10666.7 [N]
d.Qparab_xcent = 3.5 [m]


In [5]:
rax, ray, rdy = sy.symbols('rax ray rdy')

print('::: sistema de equações :::')
sumFX = rax
print(f'sumFX = {sumFX} [N]')
sumFY = ray + rdy - d.Quni_eq - d.Qparab_eq
print(f'sumFY = {printsymSF(sumFY)} [N]')
sumMA= -d.MB -d.Quni_eq*d.Quni_xcent -d.Qparab_eq*d.Qparab_xcent + rdy*d.LAD
print(f'sumMA = {printsymSF(sumMA)} [N.m]')
sol = solve([sumFX,sumFY,sumMA],[rax, ray, rdy])

print('\n::: resolução :::')
RAx = sol[rax]; print(f'RAx = {RAx} [N]')
RAy = sol[ray]; print(f'RAy = {RAy:.1f} [N]')
RDy = sol[rdy]; print(f'REy = {RDy:.1f} [N]')

::: sistema de equações :::
sumFX = rax [N]
sumFY = 1.0*ray + 1.0*rdy - 30666.667 [N]
sumMA = 3.0*rdy - 101333.333 [N.m]

::: resolução :::
RAx = 0.0 [N]
RAy = -3111.1 [N]
REy = 33777.8 [N]



- **1.1 :::** Momento devido à reação em A expresso por uma função de singularidade

\begin{equation*}
M(x) | R_A = R_{Ay} <x-0>^1  = R_{Ay} x = -3111.1 x
\end{equation*}

In [6]:
M1 = RAy*x
print(f'M1 = {printsymSF(M1)}')

M1 = -3111.111*x


- **1.2 :::** Momento devido à reação em A expresso por uma função de singularidade

\begin{equation*}
M(x) | M_B = M_B <x-L_{AB}>^0 =  4000 <x-1>^0
\end{equation*}

In [7]:
M2 = d.MB*(x-d.LAB)**0
print(f'M2 = {M2:.1f}')

M2 = 4000.0


- **1.3 :::** Momento devido a uma carga uniforme de valor $Q_C$ expressa por uma função de singularidade

\begin{equation*}
M(x)| {Q_C} = -\frac{Q_C}{2} <x-L_{AC}>^2 = -\frac{10000}{2} <x-2>^2
\end{equation*}

In [8]:
M3 = -d.QC/2*(x-d.LAC)**2
print(f'M3 = {M3}')

M3 = -20000.0*(0.5*x - 1)**2


- **1.4 :::** Momento devido a um carregamento parabólico expresso por uma função de singularidade

\begin{equation*}
\therefore\quad Q(x) = (Q_C-Q_E)(x-L_{AC})²
\end{equation*}

\begin{equation*}
M(x) | {Q_{CE}} = - \frac{Q_{CE}}{(3)(4)}<x-L_{AC}>^4 = - \frac{4000}{(3)(4)}<x-2>^4
\end{equation*}

In [9]:
M4 = - d.QC/3/4*(x-d.LAC)**4
print(f'M4 = {printsymSF(M4)}')

M4 = -13333.333*(0.5*x - 1.0)**4



- **1.5 :::** Momento devido à reação em D expresso por uma função de singularidade

\begin{equation*}
M(x) | {R_D} = R_{Dy} <x-L_{AD}>^1 = 33777.8 < x- 3>¹
\end{equation*}

In [10]:
M5 = RDy*(x-d.LAD)**1
print(f'M5 = {printsymSF(M5)}')

M5 = 33777.778*x - 101333.333


- **2 :::** Equação da linha elástica:

\begin{equation*}
\frac{\partial² v}{\partial x²} = \frac{M(x)}{EI}
\end{equation*}

com,

\begin{equation*}
M(x) = -31.11 x + 4 <x-1>^0 -\frac{10}{2} <x-2>^2 - \frac{4}{(4)(3)}<x-2>^4 + 33.8 < x- 3>¹
\quad\textrm{[kN.m]}
\end{equation*}


por integração, obtem-se

\begin{equation*}
EI\theta(x) = -31.11 \frac{x²}{2} + 4 <x-1>^1 -\frac{10}{6} <x-2>^3 - \frac{4}{60}<x-2>^5 + \frac{33.8}{2} < x- 3>² + C_1
\quad\textrm{[kN.m]}
\end{equation*}

integrando novamente resulta

\begin{equation*}
EIv(x) = -31.11 \frac{x³}{6} + \frac{4}{2} <x-1>^2 -\frac{10}{24} <x-2>^4 - \frac{4}{360}<x-2>^6 + \frac{33.8}{6} < x- 3>³ + C_1x + C_2
\quad\textrm{[kN.m]}
\end{equation*}

As constantes de integração $C_1$ e $C_2$ são determinadas pelas condições de fronteira.

\begin{equation*}
\begin{aligned}
x = 0 \quad\Rightarrow\quad y(0) = 0
&\quad\therefore\quad C_2 = 0
\end{aligned}
\end{equation*}

\begin{equation*}
\begin{aligned}
&x = L_{AE} \quad\Rightarrow\quad y(L_{AE}) = 0
\quad\therefore\quad \\
& -31.11 \frac{L_{AE}³}{6} + \frac{4}{2} <L_{AE}-1>^2 -\frac{10}{24} <L_{AE}-2>^4 - \frac{4}{360}<L_{AE}-2>^6 + \frac{33.8}{6} < L_{AE}- 3>³ + C_1L_{AE} = 0
\end{aligned}
\end{equation*}

In [11]:
c1, c2 = sy.symbols('c1 c2')

print('::: eq. cond. fronteira :::')

def flex(x):
    if x == 0:
        t1 = RAy/6*x**3
        t2, t3, t4, t5 = 0, 0, 0, 0
    else:
        t1 = RAy/6*x**3
        t2 = d.MB/2*(x-d.LAB)**2
        t3 = - d.QC/24*(x-d.LAC)**4
        t4 = - Al/360*(x-d.LAC)**6
        t5 = RDy/6*(x-d.LAD)**3
    return t1 + t2 + t3 + t4 + t5 + c1*x + c2

eqC1 = flex(0.)
sol = solve(eqC1,c2)
C2 = sol[0]
print(f'eqC2 = {eqC1}  ::  C2 = {C2}')

eqC2 = flex(d.LAD).subs(c2,C2)
sol = solve(eqC2,c1)
C1 = sol[0]
print(f'eqC2 = {printsymSF(eqC2)}  ::  C1 = {C1:.3f}')


::: eq. cond. fronteira :::
eqC2 = c2  ::  C2 = 0
eqC2 = 3.0*c1 - 6427.778  ::  C1 = 2142.593


In [12]:
def theta_(x):
    return sy.diff(flex(x).subs({(c1,C1),(c2,C2)}),x)/d.EI

def iy_(x):
    return flex(x).subs({(c1,C1),(c2,C2)})/d.EI

print(f'y(x) = {printsymSF(iy_(x))}')
print(f'y(d.LAE) = {iy_(d.LAE):.3e} [m] = {iy_(d.LAE)*1e3:.3f} [mm]')


y(x) = 0.001*x**3 - 0.008*x**2 + 0.022*x - 0.021
y(d.LAE) = -1.133e-3 [m] = -1.133 [mm]


### Apêndice: Equação da elástica

A equação $y(x)$ descreve a posição deformada do eixo da viga devido à ação do momento fletor. Esta equação é conhecida como a equação da linha elástica e pode ser obtida na prática através da solução de uma equação diferencial por dupla integração da equação do momento fletor ao longo da viga. A equação resultante deve ainda satisfazer as condições de fronteira do problema, ou seja, ser cinematicamente admissível. Esta condição permitirá determinar, nomeadamente, as constantes de integração resultantes do processo de integração.

A hipótese de Euler-Bernoulli para o problema de flexão de uma viga conduz à seguinte equação,

\begin{equation*}
\varepsilon_{xx} = -\frac{y}{\rho}
\end{equation*}

Assumindo ainda a lei de Hooke para o comportamento linear elástico do materia, $\sigma_{xx} = E \varepsilon_{xx}$, pode obter-se uma relação entre o momento flector $M(x)$ e o raio de curvatura da superfície neutra da viga $\rho$,

\begin{equation*}
\sigma_{xx} = E \varepsilon_{xx} =
-\frac{yE}{\rho}
\quad\wedge\quad
\sigma_{xx} = -\frac{My}{I_z}
\quad\therefore\quad
\frac{1}{\rho} = \frac{M}{EI_z}
\end{equation*}

Considerando agora a relação entre o raio de curvatura $\rho$ e as segundas derivadas da função $y(x)$, que para pequenos valroes resulta,

\begin{equation*}
\frac{1}{\rho} \approx \frac{\mathrm{d}² y}{\mathrm{d} x²}
\end{equation*}

obtém-se finalmente,

\begin{equation*}
\frac{\mathrm{d}² y}{\mathrm{d} x²}
= \frac{M}{EI_z}
\end{equation*}

Assim sendo, para se obter a equação da elástica da viga do problema, $y(x)$, é necessário:

1. Escrever a equação do momento fletor na viga, M(x) - nomeadamente usando funções de singularidasde;

2. Integrar duas vezes essa equação;

3. Aplicar condições fronteira para calcular as duas constantes que surgem durante a integração.

### Apêndice A: Equação da elástica

A equação $y(x)$ descreve a posição deformada do eixo da viga devido à ação do momento fletor. Esta equação é conhecida como a equação da linha elástica e pode ser obtida na prática através da solução de uma equação diferencial por dupla integração da equação do momento fletor ao longo da viga. A equação resultante deve ainda satisfazer as condições de fronteira do problema, ou seja, ser cinematicamente admissível. Esta condição permitirá determinar, nomeadamente, as constantes de integração resultantes do processo de integração.

A hipótese de Euler-Bernoulli para o problema de flexão de uma viga conduz à seguinte equação,

\begin{equation*}
\varepsilon_{xx} = -\frac{y}{\rho}
\end{equation*}

Assumindo ainda a lei de Hooke para o comportamento linear elástico do materia, $\sigma_{xx} = E \varepsilon_{xx}$, pode obter-se uma relação entre o momento flector $M(x)$ e o raio de curvatura da superfície neutra da viga $\rho$,

\begin{equation*}
\sigma_{xx} = E \varepsilon_{xx} =
-\frac{yE}{\rho}
\quad\wedge\quad
\sigma_{xx} = -\frac{My}{I_z}
\quad\therefore\quad
\frac{1}{\rho} = \frac{M}{EI_z}
\end{equation*}

Considerando agora a relação entre o raio de curvatura $\rho$ e as segundas derivadas da função $y(x)$, que para pequenos valroes resulta,

\begin{equation*}
\frac{1}{\rho} \approx \frac{\mathrm{d}² y}{\mathrm{d} x²}
\end{equation*}

obtém-se finalmente,

\begin{equation*}
\frac{\mathrm{d}² y}{\mathrm{d} x²}
= \frac{M}{EI_z}
\end{equation*}

Assim sendo, para se obter a equação da elástica da viga do problema, $y(x)$, é necessário:

1. Escrever a equação do momento fletor na viga, $M(x)$ - nomeadamente usando funções de singularidasde;

2. Integrar duas vezes essa equação;

3. Aplicar condições fronteira para calcular as duas constantes que surgem durante a integração.


### Apêndice B: Funções de singularidade

As expressões $⟨x − a⟩^0$, $⟨x − a⟩^1$, $⟨x − a⟩^2$ são designadas por funçoes singulares. Quando $n ≥ 0$,

\begin{equation*}
⟨x − a⟩^n = 
\begin{cases}
(x − a)^n & \quad\text{se } x \ge a \\
0 & \quad\text{se } x < a \\
\end{cases}
\end{equation*}

Nota-se que, os parênteses angulares ⟨ ⟩ devem ser substituídos por parênteses curvos ( ) quando $x ≥ a$ e por zero quando $x < a$.

Segue da definição de funções singularidades que,

\begin{equation*}
\int ⟨x − a⟩^n \mathrm{d}x = \frac{1}{n+1} 
⟨x − a⟩^{n+1} \quad\textrm{se}\quad n \ge 0
\end{equation*}

e 

\begin{equation*}
\frac{d}{dx} ⟨x − a⟩^n = n⟨x − a⟩^{n-1} \quad\textrm{se}\quad n \ge 1
\end{equation*}

### Anexo: Funções de singularidade

<img src="https://userweb.fct.unl.pt/~jmc.xavier/MSII/Notebooks/Formulario/singularity_functions.png"
width="900">

---

Copyright (c) DEMI - FCT NOVA

Interactive computing by <a href="https://jupyter.org/" target="_blank"> <span
style="color:#333399"> Jupyter Notebook </span> </a> &nbsp;|&nbsp;Coded by <a href = "mailto: jmc.xavier@fct.unl.pt">José Xavier</a>

Licensed under  <a href="http://creativecommons.org/licenses/by-sa/4.0/"
target="_blank"> <span style="color:#333399;font-size: 20px"> CC BY-SA 4.0  </span></a>